File: queue/js/queue.js
class Queue {
  constructor() {
    this.queue = [];
    this.front = -1;
    this.rear = -1;
  }

  isEmpty() {
    if (this.queue.length < 1) {
      return 'Empty'
    };
  };

  enqueue(value) {
    if (this.isEmpty() === 'Empty') {
      this.front += 1;
    }
    this.rear += 1;
    this.queue.push(value);
  };

  dequeue() {
    if (this.isEmpty() === 'Empty') {
      return 'Empty'
    }
    if (this.queue.length < 1) {
      this.front = -1;
      this.rear = -1;
    }
    this.queue.shift();
    this.rear -= 1;
  }

  peek() {
    return this.queue[this.front]
  }
};

newQueue = new Queue;
console.log(newQueue.peek());
newQueue.enqueue(1);
console.log(newQueue.peek());
newQueue.enqueue(2);
console.log(newQueue.peek());
newQueue.enqueue(3);
console.log(newQueue.peek());
newQueue.enqueue(4);
console.log(newQueue.peek());
newQueue.enqueue(5);
console.log(newQueue.peek());
newQueue.dequeue();
console.log(newQueue.front, newQueue.rear, newQueue.queue)
console.log(newQueue.peek());
newQueue.dequeue();
console.log(newQueue.peek());
newQueue.dequeue();
console.log(newQueue.peek());
newQueue.dequeue();
console.log(newQueue.peek());
newQueue.dequeue();
console.log(newQueue.peek());

File: queue/ruby/queue.rb
class Queue
  def initialize
    @queue = []
    @front = -1
    @rear = -1
  end

  def is_empty?
    return 'Empty' if @queue.empty?

    false
  end

  def enqueue(value)
    @front += 1 if @queue.empty?

    @rear += 1
    @queue.push(value)
  end

  def dequeue
    if @queue.empty?
      @rear = -1
      @front = -1
    end
    @queue.shift
    @rear -= 1
  end

  def peek
    @queue[@front]
  end
end

new_queue = Queue.new
puts new_queue.is_empty?
new_queue.enqueue(1)
puts new_queue.peek
new_queue.enqueue(2)
new_queue.enqueue(3)
new_queue.enqueue(4)
new_queue.enqueue(5)
new_queue.dequeue
puts new_queue.peek
new_queue.dequeue
puts new_queue.peek
new_queue.dequeue
puts new_queue.peek
new_queue.dequeue
puts new_queue.peek
new_queue.dequeue
puts new_queue.peek
